Mathlib/Topology/Connected/RealIntervals.lean

```diff
@@ -0,0 +1,560 @@
+    exact this ⟨z_le_y, LT.lt.le y_lt_x⟩
+  · push_neg at h
+    have : y ∈ lowerBounds X := fun z ↦ fun zX ↦ LT.lt.le (h z zX)
+    have : y ≤ infX := h_infX.2 this
+    linarith
+
+section inf_sup_section
+variable
+  {infX supX : ℝ}
+  (h_infX : IsGLB X infX)
+  (h_supX : IsLUB X supX)
+
+lemma connected_bdd_subset_contains_Ioo :
+    Ioo infX supX ⊆ X := by
+  let ⟨z, zX⟩ := IsConnected.nonempty conn
+  have h₁ : Ioo infX z ⊆ X := connected_bddBelow_subset_contains_Ioo conn h_infX zX
+  have h₂ : Ioo z supX ⊆ X := connected_bddAbove_subset_contains_Ioo conn h_supX zX
+  have h₃ : Ioo infX supX ⊆ Ioo infX z ∪ {z} ∪ Ioo z supX := by
+    intro x ⟨infX_lt_x, x_lt_supX⟩
+    rcases lt_trichotomy x z with (x_lt_z | x_eq_z | z_lt_z)
+    · left; left
+      exact ⟨infX_lt_x, x_lt_z⟩
+    · left; right
+      rw [mem_singleton_iff, x_eq_z]
+    · right
+      exact ⟨z_lt_z, x_lt_supX⟩
+  have h₄ : Ioo infX z ∪ {z} ∪ Ioo z supX ⊆ X := by
+    rintro x ((_ | x_eq_z) | _)
+    · apply h₁; assumption
+    · rw [mem_singleton_iff] at x_eq_z
+      rw [x_eq_z]
+      exact zX
+    · apply h₂; assumption
+  intro x x_Ioo
+  exact h₄ (h₃ x_Ioo)
+
+lemma characterize_Ioo
+    (infX_X : infX ∉ X) (supX_X : supX ∉ X) :
+    X = Ioo infX supX := by
+  apply Subset.antisymm
+  · intro x xX
+    exact ⟨excluded_infX_lt_x xX h_infX infX_X,
+           x_lt_excluded_supX xX h_supX supX_X⟩
+  · exact connected_bdd_subset_contains_Ioo conn h_infX h_supX
+
+lemma characterize_Ioc
+    (inf_lt_sup : infX < supX)
+    (infX_X : infX ∉ X) (supX_X : supX ∈ X) :
+    X = Ioc infX supX := by
+  apply Subset.antisymm
+  · intro x xX
+    exact ⟨excluded_infX_lt_x xX h_infX infX_X,
+           h_supX.1 xX⟩
+  · rw [Ioc_Ioo_with_endpoint inf_lt_sup, union_subset_iff]
+    exact ⟨connected_bdd_subset_contains_Ioo conn h_infX h_supX,
+           singleton_subset_iff.mpr supX_X⟩
+
+lemma characterize_Ico
+    (inf_lt_sup : infX < supX)
+    (infX_X : infX ∈ X) (supX_X : supX ∉ X) :
+    X = Ico infX supX := by
+  apply Subset.antisymm
+  · intro x xX
+    exact ⟨h_infX.1 xX,
+           x_lt_excluded_supX xX h_supX supX_X⟩
+  · rw [Ico_Ioo_with_endpoint inf_lt_sup, union_subset_iff]
+    exact ⟨connected_bdd_subset_contains_Ioo conn h_infX h_supX,
+           singleton_subset_iff.mpr infX_X⟩
+
+lemma characterize_Icc (infX_X : infX ∈ X) (supX_X : supX ∈ X) : X = Icc infX supX := by
+  apply Subset.antisymm
+  · exact bdd_subset_Icc h_infX h_supX
+  · exact (conn_has_Icc conn) infX infX_X supX supX_X
+
+lemma characterize_singleton (eq : infX = supX) : (X = {infX}) := by
+  rw [← eq] at h_supX
+  apply Subset.antisymm
+  · apply subset_trans (bdd_subset_Icc h_infX h_supX)
+    exact Set.Icc_subset {infX} rfl rfl
+  · rw [singleton_subset_iff]
+    rcases IsLUB.nonempty h_supX with ⟨x, xX⟩
+    have : infX = x := LE.le.antisymm (h_infX.1 xX) (h_supX.1 xX)
+    exact mem_of_eq_of_mem (id this) xX
+
+end inf_sup_section
+end conn_section
```

Comment:
Style nit: the usual convention is to leave a blank line before each section ends. (Makes it easier to add future material.) To some extend, all of this is subjective, but it's the convention mathlib has settled on.